src/facade/dragonfly_connection.cc

```diff
@@ -918,7 +922,19 @@ bool Connection::IsPrivileged() const {
 }
 
 bool Connection::IsMain() const {
-  return static_cast<Listener*>(listener())->IsMainInterface();
+  return is_main_;
+}
+
+bool Connection::HasMainListener() const {
+  if (is_main_) {
+    return true;
+  }
+
+  if (const Listener* lsnr = static_cast<Listener*>(listener()); lsnr) {
+    return lsnr->protocol() == Protocol::MEMCACHE;
+  }
+
+  return false;
```

Comment:
```suggestion
  const Listener* lsnr = static_cast<Listener*>(listener());
  return is_main_ ||  lsnr->protocol() == Protocol::MEMCACHE;
```

Reply:
and maybe rename because `HasMainListener` is kinda a lie if protocol is MEMCACHE. Maybe something like `IsMainOrMemcached` ?

Reply:
The problem with accessing `lsnr->protocol()` without a null check is that I found it to be null during tests when we end up here with `TestConnection`, and it caused segfault. From what I could see the test connection doesn't have a listener pointer.

Reply:
I made both the changes, I ordered it so the `static_cast` is only after checking for `is_main_`, but I can change it to make everything in one composite statement, in that case we always have to do the static cast.